### ISSUE TITLE
Dask easy Deadlock with gather and compute on AWS Cluster

### ISSUE THREAD

**Issue by SrSoto** [dask/distributed#7047]
On our project, we use a complex main function for processing large batches of data with type castings, transformations and file merging. The tasks structure could be summed up as follows:

1. A `main` task, including all the logical process from any of our mains that also takes care of cluster status, configs and distributed logging.
2. A 'parent' task, let's call it `step`, which aims to process one full folder of one type of our data. The `main` task calls multiple `step`'s so that we have certain workers to control part of the process.
3. A `data` task, which consists on reading, processing and writing the file given the `step` instructions. `data` sometimes calls `compute` or `persist` over `DaskDataFrame`'s that have called `map_partitions`; while `step` usually waits for all `data` tasks on a `gather`.

This is a complex process, but justified by optimal results for S3 storage on AWS and full parallelization. On local, 'one-big-machine' clusters and even on an EC2 instance (alone) works perfectly. Nevertheless, a deadlock is found between part of the `data` task (generated from a `map_partitions`) and `step`; only when executing on an AWS Cluster with different machines for Client, Scheduler and Worker roles. Seems that having a worker to dwell these two different tasks is what is causing the deadlock, but only when, by some type of 'magic', the scheduler is not on the same machine as the worker.

We, of course, expect this behavior not to happen as the code is correct on "local" cases. Currently our Data process works by omitting the `step` task, but we don't achieve full parallelization of folders being processed. As fully understanding our code and structure would make impossible this issue to be solved, a MCVE example is given imitating this dependency graph causing an interlock between `foo` and `parent_foo`.

```python
from dask.distributed import get_client, Client
client = Client("127.0.0.1:8786")
def foo():
    tmp_client = get_client()
    tmp_client.compute(None, traverse=False)
    return 0
def parent_foo():
    tmp_client = get_client()
    future = tmp_client.submit(foo, workers=[my_worker])
    return tmp_client.gather(future)
my_worker = list(client.scheduler_info()["workers"].keys())[0]
client.submit(parent_foo, workers=[my_worker])
```

This code returns the `Future` on `pending` state, while on the Dashboard we can see how these two tasks stuck on processing (as a 'green bar' for the selected worker).

The AWS Cluster architecture can be explained as follows:
1. A Bastion EC2 connects between the Client (my laptop) and the Scheduler by an SSH tunnel through `:8787` and `:8786` (that's why `"127.0.0.1:8786"` grants me access to the AWS cluster from local).
2. A Scheduler instance hosts a Docker image containing our environment and packaged code, running simply `dask-scheduler`.
3. Multiple Worker instances hosts the same Docker image, but running `dask-worker` (with some parameters as `nprocs` and the scheduler address).

**Environment**:

- Dask version: 2022.6.1
- Python version: 3.9.14.final.0
- Operating System: Amazon Linux 2
- Install method (conda, pip, source): pip

<details>
<summary>Cluster Dump State:</summary>

[dask-cluster-dump.msgpack.gz](https://github.com/dask/dask/files/9606488/dask-cluster-dump.msgpack.gz)

</details>

To sum up...

* **What is fully causing the deadlock?**
* **Why does this deadlock only happen when running on distributed remote machines?**
* **How can this scheduling or task dependencies problem be solved?**

Thanks in advance! :sunglasses:  :handshake: :sunglasses:

**Comment by crusaderky** [comment]
In the dump you provided, `my_worker` has 1 thread.
You constrained both parent_foo and foo to run exclusively on my_worker.

1. parent_foo starts on my_worker
2. it contacts the scheduler, asking it to run foo on my_worker as soon as there's a thread available
3. it then proceeds to wait until foo is terminated. But foo will never even start, because parent_foo is occupying the only worker.
4. deadlock.

Firstly, I would like to strongly question *why* you think it's necessary to launch a task from another task. Is it because `step` needs to list the contents of a folder and generate one task per file contained? In that case, it would be much better to either perform the scan from the client (if feasible and performant network-wise) or do so from a task and then return the control intermediately to the client so that more tasks can be scheduled:

```python
@delayed
def discover_folder_contents():
    return glob.glob("myfolder")  # or equivalent s3 command or whatever

@delayed
def data(fname):
    with open(fname):    # or equivalent s3 command or whatever
    ...


# This runs on the client
def main():
    client = Client(scheduler_addr)
    fnames = discover_folder_contents().compute()
    data = client.gather(client.compute([data(fname) for fname in fnames]))
```

Note however that, even with my design above, you'll get stuck again in another deadlock here:

>  data sometimes calls compute or persist over DaskDataFrame

because if you have 20 threads across the cluster and 20 data tasks running, and all of them need to call compute/persist, then there won't be anywhere to do so.

**If calling tasks from tasks is truly unavoidable** - and again, I would like you to challenge that as much as possible - then you should look into secede/rejoin: https://distributed.dask.org/en/latest/task-launch.html?highlight=secede
